### PR TITLE
can recover from overlap error in broker and coordinator by restart realtime task

### PR DESCRIPTION
**Pull request by kaijianding**
When user wants to enlarge the segment size in realtime task and rebuild from some checkpoint, like kafka offset, like from hour to day without using new version (the default versioning policy is the IntervalStartVersioningPolicy, will use same version when rebuild), it causes 'can't overlap error' in broker and coordinator when user starts the new tasks before the old segments are disabled from druid (there is some delay from disabled to removed from serverView), and 'No servers found for SegmentDescriptor'.
Even worse, if user want's to fix these error by killing the new task and restore the old config, it still fails because of 'can't overlap error' again.

It's caused by the inconsistent state of all allTimelineEntries/completePartitionsTimeline/selectors when the first time  'can't overlap error'  occurred.
The new segments are added to allTimelineEntries/completePartitionsTimeline, but not the selectors, user can't remove the new segments from allTimelineEntries/completePartitionsTimeline by killing the task due to selectors doesn't contain the new segments and refuse to clean the timelines.

We can just simply fix this issue by always trying to clean the timelines, thus user can kill the task, clean the timeline, disable the old segments and start the new task later.

### CODE REVIEW

**Comment by fjy** [comment]
@kaijianding looks good on first inspection, but how does this impact memory usage?

File: server/src/main/java/io/druid/client/BrokerServerView.java [hunk 1]
```diff
@@ -244,15 +244,15 @@ private void serverRemovedSegment(DruidServerMetadata server, DataSegment segmen
...
     String segmentId = segment.getIdentifier();
-    final ServerSelector selector;
+    ServerSelector selector;
 
     synchronized (lock) {
       log.debug("Removing segment[%s] from server[%s].", segmentId, server);
 
       selector = selectors.get(segmentId);
       if (selector == null) {
         log.warn("Told to remove non-existant segment[%s]", segmentId);
```
Comment (leventov) [inline server/src/main/java/io/druid/client/BrokerServerView.java]: Maybe should rephrase the logging statement now since and change level to "info"?

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 2]
```diff
@@ -158,10 +158,10 @@ private void serverRemovedSegment(DruidServerMetadata server, DataSegment segmen
     synchronized (lock) {
       log.debug("Removing segment[%s] from server[%s].", segmentId, server);
 
-      final SegmentLoadInfo segmentLoadInfo = segmentLoadInfos.get(segmentId);
+      SegmentLoadInfo segmentLoadInfo = segmentLoadInfos.get(segmentId);
       if (segmentLoadInfo == null) {
         log.warn("Told to remove non-existant segment[%s]", segmentId);
```
Comment (leventov) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: Maybe should rephrase the logging statement now since and change level to "info"?

**Comment by kaijianding** [comment]
Here is the detail process how this error happens.

1. run the old task with hour segmentGranularity and announce two segments: s1 ("2016-10-31T00:00:00.000+08:00/2016-10-31T01:00:00.000+08:00", "v1", NumberedShardSpec(0, 2)) and s2 ("2016-10-31T00:00:00.000+08:00/2016-10-31T01:00:00.000+08:00", "v1", NumberedShardSpec(1, 2)), and finaly handoff s1 and s3
2. user can query
3. kill the old task and disable s1 and s2 in db, but brokers will remove s1/s2 from memory serveral minutes later due to druid.coordinator.period on coordinator side and the time needed to unload s1/s2 from historicals
4. reset checkpoint and immediately run the new task to read from 2016-10-31T00:00:00.000+08:00 (we have an internal product like kafka with the ability to rewind from specified timestamp) and use day segmentGranularity
5. announce s3 (2016-10-31T00:00:00.000+08:00/2016-11-01T01:00:00.000+08:00", "v1", NumberedShardSpec(0, 2)) and s4 (2016-10-31T00:00:00.000+08:00/2016-11-01T01:00:00.000+08:00", "v1", NumberedShardSpec(1, 2))  
  5.1 in BrokerServerView.serverAddedSegment, timeline.add(...) adds s3/s4 to allTimelineEntries but failed to add to completePartitionsTimeline due to overlap error, then selectors.put(segmentId, selector) is not called and queryableDruidServer is not bound to the selector
6. administrator gets the "can't overlap error " alert
7. brokers remove s1/s2, and add s3/s4 to completePartitionsTimeline in VersionedIntervalTimeline.remove(...) 
```
    for (Map.Entry<Interval, TreeMap<VersionType, TimelineEntry>> versionEntry : allTimelineEntries.entrySet()) {
      if (versionEntry.getKey().overlap(interval) != null) {
        if (incompleteOk) {
          add(timeline, versionEntry.getKey(), versionEntry.getValue().lastEntry().getValue());
        } else {
          for (VersionType ver : versionEntry.getValue().descendingKeySet()) {
            TimelineEntry timelineEntry = versionEntry.getValue().get(ver);
            if (timelineEntry.getPartitionHolder().isComplete()) {
              add(timeline, versionEntry.getKey(), timelineEntry);
              break;
            }
          }
        }
      }
    }
```
- 8. query can find the s3/s4 when timeline.lookup() called, then "No servers found for SegmentDescriptor" raised in CachingCLusteredClient due to s3/s4 don't have any server in selector, as mentioned in step 5
- 9. administrator gets the "No servers found for SegmentDescriptor" alert, and try to kill the new task to remove s3/s4  
  9.1 In BrokerServerView.serverRemovedSegment(), because selectors don't contain s3/s4 as mentioned in step 5, and then allTimelineEntries/completePartitionsTimeline is kept no change  
  9.2 administrator gets the "No servers found for SegmentDescriptor" alert continually  
  9.3 administrator try to run old task and read from 2016-10-31T00:00:00.000+08:00 again, and s1/s2 are announced again, but failed due to overlap error (s1 conflicts with s3, s3 is still in allTimelineEntries as mentioned in step 9.1)
- 10 administrator has to restart all brokers to clean the memory and run the new task

So the broker can't recover from "overlap error" by itself without restart.
To fix this issue simply, we can clean the memory state at step 9.1 to remove s3/s4 from allTimelineEntries/completePartitionsTimeline, and run the new task several minutes later to announce s3/s4 again, at that time the announcement will success

Also we can ensure always to use new version by using CustomVersioningPolicy to avoid overlap error, but it's better we also fix this issue in broker with the ability to clean the memory state by removing the overlapped segments

File: server/src/main/java/io/druid/client/BrokerServerView.java (hunk 1 above)
Comment (kaijianding) [inline server/src/main/java/io/druid/client/BrokerServerView.java]: IMO, this case is still abnormal and better keep it as warn level.

**Comment by fjy** [comment]
👍

**Comment by himanshug** [comment]
@kaijianding looks like fundamental problem is that broker side of things made assumption that different segment sets would have different version which appears to be broken now.
are you hitting it with new kafka-indexing-service only? 
@gianm @dclim can you confirm if the behavior of using interval start-time as segment version is intentional ?

**Comment by kaijianding** [comment]
@himanshug I'm not hitting it with the new kafka-indexing-service.
We have an internal product which is similar to kafka, but with ability to rewind from old timestamp(kafka can only rewind from offset which is difficult to associate offset with a specified time point).
We have written a druid extension to read from it and remember the checkpoint after segment handoff.
Sometimes, we want to enlarge segmentGranularity to reduce segments number to lessen broker memory pressure, then we rewind from old time and rebuild segments (usually because we found the segmentGranularity is not properly configed).
The best way to resolve this issue is that we use CustomVersioningPolicy to always pass new timestamp as new version in our extension.
But we still think it is worth making broker can recover from "overlap error" by killing the new task, otherwise, all brokers are down in case someone else using druid hits this issue by mistake.

Basically, there are some preconditions to hit this issue:
1. enlarge segmentGranularity, like from hour to day
2. using the default IntervalStartVersioningPolicy
3. can rewind from old data timestamp to create the same interval start as version, like 2016-10-31T00:00:00.000+08:00

**Comment by himanshug** [comment]
@kaijianding i see, thanks for the detailed explanation. actually, to the best of my knowledge, none of the druid ingestions use interval start time as segment version and new segments always have higher versions. i'm not sure what else breaks down when that assumption is taken away.

That said, IntervalStartVersioningPolicy does exist in the code and I'm not sure when it would be used. if we do want to make the update then, I think, VersionedIntervalTimeline should be modified to handle the case of overlapping segments with same version.

does anyone else have an opinion ?

**Comment by kaijianding** [comment]
So, do we still need the fix in this PR, or waiting for VersionedIntervalTimeline be modified, or it shouldn't be a real problem therefore no fix needed? @himanshug

**Comment by himanshug** [comment]
@kaijianding let me discuss this in the dev meeting, will update the PR afterwards.

**Comment by himanshug** [comment]
@kaijianding it was discussed among the devs and we believe broker making the assumption about "different segments sets having different versions, and it should fail otherwise" is there by design and does not need to be changed/fixed.

however, in the long terms, we might want to make updates to handle the situation more gracefully and not getting into a corrupted state after above failure where broker stops accepting queries altogether.

**Comment by himanshug** [comment]
based on the dev-sync discussion, I'm closing this PR.
